unittests/runtime-io/io.cpp

```diff
@@ -0,0 +1,343 @@
+  int fd = overwriteTestFile();
+  ::close(fd);
+
+  auto realFilename = makeString("test.txt");
+  auto fakeFilename = makeString("testFake.txt");
+  auto control = makeString("r");
+
+  block * b1 = hook_IO_open(realFilename, control);
+  BOOST_CHECK(0 < mpz_cmp_si((mpz_ptr) *(b1->children), 0));
+
+  block * b2 = hook_IO_open(fakeFilename, control);
+  BOOST_CHECK_EQUAL((uint64_t)*(b2->children), ERRBLOCK(getTagForSymbolName(ERRTAG(ENOENT))));
+}
+
+BOOST_AUTO_TEST_CASE(tell) {
+  mpz_t f;
+  int fd = overwriteTestFile();
+  mpz_init_set_si(f, fd);
+
+  block * b = hook_IO_tell(f);
+
+  BOOST_CHECK_EQUAL(0, mpz_cmp_si((mpz_ptr) *(b->children), lseek(mpz_get_si(f), 0, SEEK_CUR)));
+
+  lseek(mpz_get_si(f), 5, SEEK_CUR);
+
+  b = hook_IO_tell(f);
+  BOOST_CHECK_EQUAL(0, mpz_cmp_si((mpz_ptr) *(b->children), 5));
+
+  b = hook_IO_tell(f);
+  BOOST_CHECK_EQUAL(0, mpz_cmp_si((mpz_ptr) *(b->children), lseek(mpz_get_si(f), 0, SEEK_CUR)));
+
+  lseek(mpz_get_si(f), -4, SEEK_CUR);
+
+  b = hook_IO_tell(f);
+  BOOST_CHECK_EQUAL(0, mpz_cmp_si((mpz_ptr) *(b->children), 1));
+
+  b = hook_IO_tell(f);
+  BOOST_CHECK_EQUAL(0, mpz_cmp_si((mpz_ptr) *(b->children), lseek(mpz_get_si(f), 0, SEEK_CUR)));
+
+  ::close(fd);
+
+  mpz_set_si(f, -1);
+  b = hook_IO_tell(f);
+  BOOST_CHECK_EQUAL((uint64_t)*(b->children), ERRBLOCK(getTagForSymbolName(ERRTAG(EBADF))));
+}
+
+BOOST_AUTO_TEST_CASE(getc) {
+  mpz_t f;
+  int fd = overwriteTestFile();
+  mpz_init_set_si(f, fd);
+
+  block * b = hook_IO_getc(f);
+  BOOST_CHECK_EQUAL(0, mpz_cmp_si((mpz_ptr) *(b->children), int('h')));
+
+  b = hook_IO_getc(f);
+  BOOST_CHECK_EQUAL(0, mpz_cmp_si((mpz_ptr) *(b->children), int('e')));
+
+  b = hook_IO_getc(f);
+  BOOST_CHECK_EQUAL(0, mpz_cmp_si((mpz_ptr) *(b->children), int('l')));
+
+  b = hook_IO_getc(f);
+  BOOST_CHECK_EQUAL(0, mpz_cmp_si((mpz_ptr) *(b->children), int('l')));
+
+  b = hook_IO_getc(f);
+  BOOST_CHECK_EQUAL(0, mpz_cmp_si((mpz_ptr) *(b->children), int('o')));
+
+  ::close(fd);
+
+  mpz_set_si(f, -1);
+  b = hook_IO_getc(f);
+  BOOST_CHECK_EQUAL((uint64_t)*(b->children), ERRBLOCK(getTagForSymbolName(ERRTAG(EBADF))));
+}
+
+BOOST_AUTO_TEST_CASE(read) {
+  mpz_t f;
+  mpz_t len;
+  int fd = overwriteTestFile();
+  mpz_init_set_si(f, fd);
+  mpz_init_set_si(len, 6);
+
+  block * b = hook_IO_read(f, len);
+  string * str = (string *) *(b->children);
+
+  BOOST_CHECK_EQUAL(0, strncmp(str->data, "hello ", 6));
+
+  b = hook_IO_read(f, len);
+  str = (string *) *(b->children);
+  BOOST_CHECK_EQUAL(0, strncmp(str->data, "world!", 6));
+
+  ::close(fd);
+
+  b = hook_IO_read(f, len);
+  BOOST_CHECK_EQUAL((uint64_t)*(b->children), ERRBLOCK(getTagForSymbolName(ERRTAG(EBADF))));
+}
+
+BOOST_AUTO_TEST_CASE(close) {
+  mpz_t f1;
+  mpz_t f2;
+  int fd1 = overwriteTestFile();
+  mpz_init_set_si(f1, fd1);
+  int fd2 = overwriteTestFile();
+  mpz_init_set_si(f2, fd2);
+
+  hook_IO_close(f1);
+  hook_IO_close(f2);
+
+  BOOST_CHECK_EQUAL(-1, fcntl(fd1, F_GETFD));
```

Comment:
can we check that errno == EBADF too? otherwise this test isn't really testing that the file descriptor has been closed.